File: src/shared/subagents.ts
// First-class subagent nodes shared by host and webview: one summary per delegated child, whatever the wire dialect

import type { PermissionBlock, QuestionBlock, Turn } from './transcript';

export type SubagentState = 'running' | 'completed' | 'failed' | 'cancelled' | 'disconnected';

// What Acpira can observe of the child: 'session' = the agent streams the child's own updates under its own session id
// (claude-agent-acp native sessions / RFD #1992); 'nested' = the child's tool calls arrive on the parent session with a
// structured parent link (Devin `cognition.ai/subagent_context`, Claude `claudeCode.parentToolUseId`); 'receipt' = only the
// delegation call and its result text (Kimi's Agent tool)
export type SubagentVisibility = 'session' | 'nested' | 'receipt';

export interface SubagentSummary {
  id: string;                 // Acpira's own stable id (randomUUID) — never a peer id
  parentId?: string;          // another summary's id; absent = child of the root session
  turnIndex: number;          // index in SessionView.turns of the root agent turn during which it was announced
  visibility: SubagentVisibility;
  title?: string;             // short label (RFD/claude `name`, Devin `title`, Claude legacy/Kimi `description`)
  task?: string;              // the delegated task text (RFD `task`, Devin `task`, Claude/Kimi `prompt`)
  role?: string;              // Devin `profile`, Claude/Kimi `subagent_type`; absent when not structured
  state: SubagentState;
  stateSource: 'agent' | 'local';   // 'local' = Acpira synthesized it (disconnected on connection end / parent prompt returned first)
  controls: { cancel: boolean };    // default false; only true when the agent said so for this child
  cancelRequested?: boolean;
  background?: boolean;       // the agent said the child runs detached (Devin isBackground, Claude toolResponse.isAsync)
  announcedAt: number;
  endedAt?: number;
  model?: string;             // only when reported (Devin subagent_started.model, Claude toolResponse.resolvedModel / rawInput.model)
  usage?: { used: number; size: number };  // only when a usage_update was attributed to this child
  peer: { sessionId?: string; agentId?: string; toolCallId?: string };  // the agent's own identifiers (diagnostics + replay re-association)
  activity?: string;          // latest activity label maintained host-side
  toolCount: number;
  result?: string;            // final result text the parent received (receipt content / Devin summary); for 'session' visibility the child's own last text
  permissions?: PermissionBlock[];  // this child's pending permission cards, mirrored so the root view can show them with provenance
  question?: QuestionBlock;         // this child's open question card, same reason
}

export interface SubagentRecord extends Omit<SubagentSummary, 'permissions' | 'question'> {
  turns: Turn[];
  rev?: number;   // the node's dirty counter at write time; absent on records written before it was persisted
}

File: rust/crates/acpira-shared/src/subagents.rs
//! First-class subagent nodes shared by host and webview (mirror of src/shared/subagents.ts)

use serde::{Deserialize, Serialize};

use crate::num::Num;
use crate::transcript::{PermissionBlock, QuestionBlock, Turn};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SubagentState {
  Running,
  Completed,
  Failed,
  Cancelled,
  Disconnected,
}

impl SubagentState {
  pub fn is_terminal(self) -> bool {
    !matches!(self, SubagentState::Running)
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SubagentVisibility {
  Session,
  Nested,
  Receipt,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StateSource {
  Agent,
  Local,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct SubagentControls {
  pub cancel: bool,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentPeer {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub session_id: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub agent_id: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub tool_call_id: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct SubagentUsage {
  pub used: Num,
  pub size: Num,
}

/// Fields common to the summary pushed to the webview and the persisted record
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentCore {
  pub id: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub parent_id: Option<String>,
  #[serde(deserialize_with = "crate::num::lenient_u64")]
  pub turn_index: u64,
  pub visibility: SubagentVisibility,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub title: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub task: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub role: Option<String>,
  pub state: SubagentState,
  pub state_source: StateSource,
  #[serde(default)]
  pub controls: SubagentControls,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub cancel_requested: Option<bool>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub background: Option<bool>,
  #[serde(deserialize_with = "crate::num::lenient_i64")]
  pub announced_at: i64,
  #[serde(default, skip_serializing_if = "Option::is_none", deserialize_with = "crate::num::lenient_opt_i64")]
  pub ended_at: Option<i64>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub model: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub usage: Option<SubagentUsage>,
  #[serde(default)]
  pub peer: SubagentPeer,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub activity: Option<String>,
  #[serde(default, deserialize_with = "crate::num::lenient_u64")]
  pub tool_count: u64,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub result: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SubagentSummary {
  #[serde(flatten)]
  pub core: SubagentCore,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub permissions: Option<Vec<PermissionBlock>>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub question: Option<QuestionBlock>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SubagentRecord {
  #[serde(flatten)]
  pub core: SubagentCore,
  #[serde(default)]
  pub turns: Vec<Turn>,
  #[serde(default, skip_serializing_if = "Option::is_none", deserialize_with = "crate::num::lenient_opt_i64")]
  pub rev: Option<i64>,
}
